### ISSUE TITLE
performance regression in winnow 0.4.0

### ISSUE THREAD

**Issue by martinohmann** [winnow-rs/winnow#223]
### Please complete the following tasks

- [X] I have searched the [discussions](https://github.com/winnow-rs/winnow/discussions)
- [X] I have searched the [open](https://github.com/winnow-rs/winnow/issues) and [rejected](https://github.com/winnow-rs/winnow/issues?q=is%3Aissue+label%3AS-wont-fix+is%3Aclosed) issues

### rust version

rustc 1.68.0 (2c8cc3432 2023-03-06)

### winnow version

v0.4.0

### Minimal reproducible code

`Cargo.toml`:

```toml
[package]
name = "winnow-regression"
version = "0.1.0"
edition = "2021"

[[bench]]
name = "bench"
harness = false

[dev-dependencies]
criterion = "0.4.0"
winnow = "0.3.6" # Change to 0.4.0 to see the regression in the benchmark
```

`benches/bench.rs`:

```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion, Throughput};
use std::time::Duration;
use winnow::{character::multispace0, error::Error, Parser};

fn bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("parse");

    let input = "\t ".repeat(1024 * 1024);

    group.throughput(Throughput::Bytes(input.len() as u64));
    group.warm_up_time(Duration::from_secs(5));
    group.measurement_time(Duration::from_secs(10));

    group.bench_function("multispace0", |b| {
        multispace0::<&str, Error<&str>>.parse_next(&input).unwrap();
        b.iter(|| black_box(multispace0::<&str, Error<&str>>.parse_next(&input).unwrap()))
    });

    group.finish();
}

criterion_group!(benches, bench);
criterion_main!(benches);
```

### Steps to reproduce the bug with the above code

Run `cargo criterion` using `winnow` 0.3.6 and 0.4.0 to see the performance difference.

### Actual Behaviour

Benchmark result with winnow 0.4.0:

```text
parse/multispace0       time:   [2.4784 ms 2.4805 ms 2.4831 ms]
                        thrpt:  [805.45 MiB/s 806.29 MiB/s 806.97 MiB/s]
                 change:
                        time:   [+45.529% +45.820% +46.096%] (p = 0.00 < 0.05)
                        thrpt:  [-31.552% -31.422% -31.285%]
                        Performance has regressed.
```

### Expected Behaviour

Benchmark result with winnow 0.3.6:

```
parse/multispace0       time:   [1.7021 ms 1.7058 ms 1.7113 ms]
                        thrpt:  [1.1413 GiB/s 1.1450 GiB/s 1.1475 GiB/s]
                 change:
                        time:   [-0.6179% -0.4022% -0.2125%] (p = 0.00 < 0.05)
                        thrpt:  [+0.2129% +0.4038% +0.6217%]
                        Change within noise threshold.
```

### Additional Context

Hi there!

I'm maintaining https://github.com/martinohmann/hcl-rs, a library to parse and encode HCL and am currently working on adding editing capabilities, which requires a complete rewrite of the parser. The rewrite largely borrows ideas from `toml-edit` (thanks for this awesome piece!), but i didn't release yet as it is still in prototyping phase (see the `hcl-edit` branch if you're curious https://github.com/martinohmann/hcl-rs/tree/hcl-edit/crates/hcl-edit).

I hit a ~15-30% performance regression in my parser (depending on the input) when upgrading winnow to 0.4.0 and did a `git bisect` to pin it down to https://github.com/winnow-rs/winnow/commit/5ff0e8ccbef7cd3289658fd8f9627cb6a4274401 which was merged via https://github.com/winnow-rs/winnow/pull/210.

To me it looks like the refactor introduced some inefficiencies in various combinators mainly caused by the merge of complete/streaming logic. Please note, while the reproducer above uses `multispace0`, other combinators are affected as well.

**Comment by epage** [comment]
Thank you for the detailed issue and the work you've put into it!  The complete/streaming refactors were a bunch of performance landmines.  I held off on the ones that I found and some I still need to dig into to understand why they slowed things down.

Even this case is a good example.  In theory, the generated code should be the same between these. 

- `(' ', '\t', '\r', '\n')` should be the same as
- `' '.contains_token(c) || '\t'.contains_token(c) || '\r'.contains_token(c) || '\n'.contains_token(c)` which would be the same as
- `' ' == c.as_char() || '\t' == c.as_char() || '\r' == c.as_char() || '\n' == c.as_char()` which would be the same as
- `let c = c.as_char(); ' ' == c || '\t' == c || '\r' == c || '\n' == c` which would be the same as

I'm going to play around with it a little bit to see if something else is going on.  We should probably get micro benchmarks for the different forms of `ContainsToken` so we have good guidance on what should be used in hot loops.

**Comment by epage** [comment]
Interesting to see the results of #225
```
contains_token/str/canada
                        time:   [35.200 ms 35.888 ms 36.630 ms]
                        thrpt:  [58.608 MiB/s 59.818 MiB/s 60.987 MiB/s]
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild
contains_token/slice/canada
                        time:   [3.6746 ms 3.7295 ms 3.7903 ms]
                        thrpt:  [566.39 MiB/s 575.62 MiB/s 584.21 MiB/s]
Found 13 outliers among 100 measurements (13.00%)
  6 (6.00%) high mild
  7 (7.00%) high severe
contains_token/tuple/canada
                        time:   [3.8404 ms 3.9019 ms 3.9690 ms]
                        thrpt:  [540.88 MiB/s 550.18 MiB/s 558.99 MiB/s]
Found 9 outliers among 100 measurements (9.00%)
  8 (8.00%) high mild
  1 (1.00%) high severe
contains_token/closure-or/canada
                        time:   [4.0458 ms 4.1101 ms 4.1760 ms]
                        thrpt:  [514.07 MiB/s 522.32 MiB/s 530.62 MiB/s]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
contains_token/closure-matches/canada
                        time:   [4.0378 ms 4.1045 ms 4.1736 ms]
                        thrpt:  [514.37 MiB/s 523.03 MiB/s 531.66 MiB/s]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```

**Comment by epage** [comment]
To add even more confusion to this
- Adding `black_box` to the inputs and outputs in the above benchmark flipped things and made `tuple` faster than `closure-or` and `closure-matches`
- Adding `#{inline(never)]` to the functions under test flipped things again and made `tuple` faster than `slice`
- With `#[inline(never)]`, I was able to more directly compare the assembly and the assembly for `tuple` was the same as `closure-or`

**Comment by epage** [comment]
quick summary of my playing with the Issue's multispace0 benchmark
- Changing to an array pattern had no performance impact
- Changing to an array reference pattern got us halfway to the closure
- Changing to a slice pattern had no performance impact
- Removing the `trace` had no performance impact

**Comment by epage** [comment]
Since the issue / discussion was foxused on multispace, closing this out due to #224.  Feel free to create other issues as you identify other regressions that our testing missied.

**Comment by epage** [comment]
(and 0.4.1 is out with the fix)

**Comment by martinohmann** [comment]
@epage Thanks for the quick fix! I just ran my own parser's benchmarks again and this is what I got:

- winnow 0.4.0 is ~15-30% slower than 0.3.6
- winnow 0.4.1 is ~5-10% slower than 0.3.6

So there's still a performance regression, but it's not as big anymore. I'll try to dig a bit more and may open another issue when I found the potential source for this.

In general I'm very happy with winnow's performance. My current HCL parser is alreay more than 100% faster than the old pest parser and it even keeps track of every single whitespace and spans, which the old parser didn't do :wink:.

**Comment by martinohmann** [comment]
@epage Yeah, I took over quite some of the great ideas you had for `toml_edit`. I initially started out with a rewrite in `nom` and hit some of the same issues you mentioned in their issue tracker. I found your `nom8` fork and was excited when you converted it into `winnow`, which was the point where I migrated immediately. This allowed me to get rid of quite some custom combinators and types that I wrote because of shortcomings of `nom` and `nom_locate`.

The despanning was a great idea to help performance and I switched to it after the first iteration of my parser had performance issues because of all the whitespace and comments, and my long term goal still is to also use the new parser for `hcl-rs`. So performance needs to be at least as good as the old `pest` parser.

I'll eventually summarize some of the learnings I had until now in a discussion, because there are also some areas were `winnow` falls short, but I guess these could be fixed in the long term, e.g. large input and output types cause massive performance degradation, which is one of the reasons I also use state in certain areas (the `Expression` enum I use for representing HCL expressions is almost 200bytes and cannot get much smaller without excessive boxing, which harms both UX and performance).

Regarding the error type I'll add a comment to #103.
